### PR TITLE
More code cleanup

### PR DESCRIPTION
**Pull request by drwill-ms**
More hub service code cleanup

### CODE REVIEW

File: e2e/test/iothub/SasCredentialAuthenticationTests.cs
Comment (drwill-ms) [inline e2e/test/iothub/SasCredentialAuthenticationTests.cs:101]: .NET track2 standards for naming date times are to use the keyword **on**.

File: e2e/test/iothub/service/ExportDevicesTests.cs
Comment (drwill-ms) [inline e2e/test/iothub/service/ExportDevicesTests.cs:202]: progress seems to be only 0 or 100, so removing from the API surface so the user doesn't look to use it. Instead, they should use status. If progress ever gets interesting, we can make it public.

File: iothub/service/src/Amqp/ClientWebSocketTransport.cs
Comment (drwill-ms) [inline iothub/service/src/Amqp/ClientWebSocketTransport.cs:61]: Deprecated Fx assertions, which were just calling Debug.Assert anyway.

File: iothub/service/src/Configurations/ConfigurationsClient.cs
Comment (drwill-ms) [inline iothub/service/src/Configurations/ConfigurationsClient.cs:22]: fields order:
1. const
2. static readonly
3. static
4. readonly
5. instance

File: iothub/service/src/Configurations/ConfigurationsClient.cs
Comment (drwill-ms) [inline iothub/service/src/Configurations/ConfigurationsClient.cs:123]: As mentioned before, check preconditions outside of the try/catch.

File: iothub/service/src/Configurations/Models/Configuration.cs
Comment (drwill-ms) [inline iothub/service/src/Configurations/Models/Configuration.cs:45]: Initialize inline

File: iothub/service/src/DigitalTwin/DigitalTwinsClient.cs
Comment (drwill-ms) [inline iothub/service/src/DigitalTwin/DigitalTwinsClient.cs:364]: No reason to call format

File: iothub/service/src/Feedback/MessageFeedbackProcessorClient.cs
Comment (drwill-ms) [inline iothub/service/src/Feedback/MessageFeedbackProcessorClient.cs:249]: Readability

File: iothub/service/src/Http/HttpClientFactory.cs
Comment (drwill-ms) [inline iothub/service/src/Http/HttpClientFactory.cs:33]: Tim, this is still used in service point manager.

File: iothub/service/src/Http/HttpClientFactory.cs
Comment (drwill-ms) [inline iothub/service/src/Http/HttpClientFactory.cs:75]: Tim, down here.

Comment (drwill-ms) [reply iothub/service/src/Http/HttpClientFactory.cs:75]: Will this go away?

File: iothub/service/src/Http/HttpMessageHelper.cs
Comment (drwill-ms) [inline iothub/service/src/Http/HttpMessageHelper.cs:123]: eTag was already a string.

File: iothub/service/src/Jobs/Models/DeviceJobStatistics.cs
Comment (drwill-ms) [inline iothub/service/src/Jobs/Models/DeviceJobStatistics.cs:11]: Moved to its own file.

File: iothub/service/src/Jobs/Models/JobQueryOptions.cs
Comment (drwill-ms) [inline iothub/service/src/Jobs/Models/JobQueryOptions.cs:9]: Inherits from the base and adds 2 more optional values, so we go from 3 to 1 params.

File: iothub/service/src/Jobs/Models/ScheduledJob.cs
Comment (drwill-ms) [inline iothub/service/src/Jobs/Models/ScheduledJob.cs:96]: Present properties with TimeSpan, instead of seconds. Serializable property is now internal (just below).

File: iothub/service/src/Query/QueryClient.cs
Comment (drwill-ms) [inline iothub/service/src/Query/QueryClient.cs:23]: Cache URIs for non-dynamic paths.

File: iothub/service/src/Utilities/Fx.cs
Comment (drwill-ms) [inline iothub/service/src/Utilities/Fx.cs:10]: Moved from another dir, with everything but this method removed.

@brycewang-microsoft, this should be on your list of things to update for the exceptions work.

File: iothub/service/src/Configurations/ConfigurationsClient.cs
Comment (timtay-microsoft) [inline iothub/service/src/Configurations/ConfigurationsClient.cs:324]: We talked about removing this validation, so you can remove this

File: iothub/service/src/Http/HttpClientFactory.cs
Comment (timtay-microsoft) [reply iothub/service/src/Http/HttpClientFactory.cs:33]: Yes, that is by design. We use the service point manager to set this timeout so that users avoid problems around failover scenarios. Did you want to change this in some way?

File: iothub/service/src/Http/HttpClientFactory.cs
Comment (timtay-microsoft) [reply iothub/service/src/Http/HttpClientFactory.cs:75]: No, this was the fix we provided so that a single http client instance would eventually start making requests to the correct endpoint if failover occurs.

Comment (drwill-ms) [reply iothub/service/src/Http/HttpClientFactory.cs:75]: I thought we were going to let customers give us an HttpClient instance where this would already have been set. But I know we prefer to have smart defaults, so if we're capable of doing this, then I'm fine.

Comment (drwill-ms) [reply iothub/service/src/Http/HttpClientFactory.cs:75]: Obviously, this one has nothing to do with the HttpClient instance, we just set it in the same method. I thought there was some problem netstandard2.0 where we couldn't set this value, though. Memory is hazy.

File: iothub/service/src/Registry/DevicesClient.cs
Comment (drwill-ms) [reply iothub/service/src/Registry/DevicesClient.cs:913]: It is unreferenced.

Comment (drwill-ms) [reply iothub/service/src/Registry/DevicesClient.cs]: The original functions for starting an import/export job used this path: /$admin/{0} where {0} is exportRegistry or importRegistry. I can't find that URI in the swagger.



For some reason, the newer API call (the one I added) uses what is in the swagger of /jobs/create.

File: iothub/service/src/Http/HttpClientFactory.cs
Comment (timtay-microsoft) [reply iothub/service/src/Http/HttpClientFactory.cs:75]: > I thought we were going to let customers give us an HttpClient instance where this would already have been set. But I know we prefer to have smart defaults, so if we're capable of doing this, then I'm fine.

This setting fell under the category of a "smart default" last we talked about it. We agreed that our client's should be resilient to failover, and this same setting is applied to the default http client in track 2

> Obviously, this one has nothing to do with the HttpClient instance, we just set it in the same method. I thought there was some problem netstandard2.0 where we couldn't set this value, though. Memory is hazy.

The problem was that we had different HttpClientHandlers depending on the .NET target framework. The track 2 logic [here](https://github.com/Azure/azure-sdk-for-net/blob/cc96e605365ebfd30f9dd00a6dff47f89feb82d8/sdk/core/Azure.Core/src/Pipeline/ServicePointHelpers.cs#L87) demonstrates the problem.

**Review by timtay-microsoft — APPROVED** [review]

File: iothub/service/src/Registry/DevicesClient.cs
Comment (drwill-ms) [inline iothub/service/src/Registry/DevicesClient.cs:951]: @timtay-microsoft @abhipsaMisra what do you think about removing this (and delete if match etag) validation is removed (like we have for non-bulk operations)?

To me, the downside here is that they won't get immediate errors, rather they have to check the errors blob, which is much more costly.

File: iothub/service/src/Messaging/MessagingClient.cs
Comment (patilsnr) [inline iothub/service/src/Messaging/MessagingClient.cs:65]: When would we uncomment this code? Could maybe benefit with an explanation of that scenario -- kind of subjective though.

File: iothub/service/src/Feedback/Models/FeedbackRecord.cs
Comment (drwill-ms) [reply iothub/service/src/Feedback/Models/FeedbackRecord.cs:46]: Not a nit. Thanks for catching this.

File: iothub/service/src/Registry/Models/JobProperties.cs
Comment (abhipsaMisra) [inline iothub/service/src/Registry/Models/JobProperties.cs:145]: 👍🏽

**Review by abhipsaMisra — APPROVED** [review]

File: iothub/service/src/Registry/DevicesClient.cs
Comment (timtay-microsoft) [reply iothub/service/src/Registry/DevicesClient.cs:951]: I'm fine with keeping this validation here for the reasons you outlined above. In the non-bulk APIs, it is easy enough for the customer to handle a 412

Comment (abhipsaMisra) [reply iothub/service/src/Registry/DevicesClient.cs:951]: +1 to that. No point in increasing an additional out-of-the-way step for the user to discover the error.

File: iothub/service/src/Registry/Models/JobProperties.cs
Comment (drwill-ms) [inline iothub/service/src/Registry/Models/JobProperties.cs:42]: Instead of factory methods, we have 2 ctors, one for import and one for export. Comments indicate which are for which, as well as what additional properties are relevant, but optional.

File: iothub/service/src/Registry/Models/JobProperties.cs
Comment (drwill-ms) [inline iothub/service/src/Registry/Models/JobProperties.cs:128]: Works as a URI.

File: e2e/test/iothub/service/ImportDevicesTests.cs
Comment (abhipsaMisra) [inline e2e/test/iothub/service/ImportDevicesTests.cs:193]: we're passing in containerUri twice to the constructor?

Comment (abhipsaMisra) [reply e2e/test/iothub/service/ImportDevicesTests.cs:193]: ah, I see your comment below about these being input and output container URIs.

File: iothub/service/src/Registry/Models/JobProperties.cs
Comment (abhipsaMisra) [inline iothub/service/src/Registry/Models/JobProperties.cs:42]: no params for doc comments?

Comment (drwill-ms) [reply iothub/service/src/Registry/Models/JobProperties.cs:42]: Adding, thanks.

File: iothub/service/src/Registry/DevicesClient.cs
Comment (drwill-ms) [inline iothub/service/src/Registry/DevicesClient.cs:229]: This call fails if I don't include this.

Interestingly enough, I can't pass null, as ETag is not nullable. So the code in this method which checks for null is wrong (or we need to make the param nullable.

Thoughts @timtay-microsoft ?

File: e2e/test/iothub/service/MessageFeedbackReceiverE2ETest.cs
Comment (drwill-ms) [inline e2e/test/iothub/service/MessageFeedbackReceiverE2ETest.cs:54]: Rather than looping, we can just await a task completion source.

File: e2e/test/iothub/service/MessageFeedbackReceiverE2ETest.cs
Comment (drwill-ms) [inline e2e/test/iothub/service/MessageFeedbackReceiverE2ETest.cs:49]: Validate we got the message that was sent.

File: iothub/service/src/Feedback/MessageFeedbackProcessorClient.cs
Comment (drwill-ms) [inline iothub/service/src/Feedback/MessageFeedbackProcessorClient.cs:199]: DateTimeOffset was not supported, so I rolled that back.

**Review by patilsnr — APPROVED** [review]
Everything looks fine to me! Also glad that Tim took a look though -- maybe wait for another's approval as well.

File: iothub/service/src/Registry/DevicesClient.cs
Comment (timtay-microsoft) [reply iothub/service/src/Registry/DevicesClient.cs:229]: For this API, there is never an ETag since the user is only giving us the device Id. There is no reason to call ```HttpMessageHelper.ConditionallyInsertETag``` because of that.

Comment (timtay-microsoft) [reply iothub/service/src/Registry/DevicesClient.cs:229]: Why are you replacing what was here? Calling the overload with a new ```Device``` instance with an empty etag seemed fine to me

Comment (drwill-ms) [reply iothub/service/src/Registry/DevicesClient.cs:229]: If I don't make that call, it fails.

Comment (drwill-ms) [reply iothub/service/src/Registry/DevicesClient.cs:229]: Because the device one ran through extra code this overload didn't need.

Comment (drwill-ms) [reply iothub/service/src/Registry/DevicesClient.cs:229]: Except, apparently, adding the ETag.

**Pull request merged**
